### ISSUE TITLE
Any thing else to do after migrating duke to Java 17?

### ISSUE THREAD

**Issue by damithc** [nus-oss/duke-J17#1]
After checking the three branches for any problems (or things that need updating), reply here to share what you did and if you found anything needing attention.

When testing the master branch, <mark>remember to update `build.gradle` to point to the correct Main class first.</mark>

@baskargopinath, @bennyLCK, @billyhoce, @gongg21, @hjungwoo01,
@jedkohjk, @jyue487, @itstrueitstrueitsrealitsreal, @brein62,
@aureliony, @ljy0422, @macareonie, @ryanozx, @Carlintyj,
@teoks0199, @drustanyjt

**Comment by itstrueitstrueitsrealitsreal** [comment]
This might not be related to the migration, but while trying to build and run the project to test if there were any breaking changes, I noticed that `build.gradle` in the `add-gradle-support`  had this line of code: 

```
mainClass.set("seedu.duke.Duke")
```

While trying to build the project, I encountered this error:
<img width="922" alt="Screenshot 2024-07-14 at 8 12 50 PM" src="https://github.com/user-attachments/assets/b37c67ef-e7c8-4d7f-8d05-e327f1da5f4d">

While trying to run the generated `.jar` file, I encountered this error:
<img width="573" alt="Screenshot 2024-07-14 at 8 13 47 PM" src="https://github.com/user-attachments/assets/eeab3fb4-57ca-4209-9828-35988291e026">

These errors imply that the incorrect class path is set in `build.gradle`, so I would suggest that we instead set the main class as:
```
mainClass.set("Duke")
```

in `build.gradle` to avoid confusion in the future, like in https://github.com/nus-cs2103-AY2324S2/forum/issues/127, https://github.com/nus-cs2103-AY2324S2/forum/issues/148, https://github.com/nus-cs2103-AY2324S2/forum/issues/296.

As changing the main class is mentioned in https://se-education.org/guides/tutorials/gradle.html, this may already be covered, so I'll leave the decision to change the main class name to the teaching team, as the main class name might differ from project to project.

However, I would suggest changing it to `Duke`, as might allow it to work with the current iteration of the project, and help students recognise how setting the main class links to the main class of their project better, instead of leaving the current `add-gradle-support` incompatible with the vanilla version of Duke students are given.

Otherwise, I didn't encounter any other issues.

**Comment by macareonie** [comment]
OS used: Windows 11
JDK distribution used: Oracle OpenJDK version 17.0.11

Built and attempted to run the project on all 3 branches. `master` and `full-template` branches had no issues. 
I encountered the same error in the `add-gradle-support` branch as @itstrueitstrueitsrealitsreal due to the incorrect main class name in `build.gradle`:


![image](https://github.com/user-attachments/assets/60050bf1-ca77-49f3-9ee1-ff68d9e5b679)

Likewise, I agree with him in changing it to `Duke` so that it is compatible with the current project iteration.

Otherwise, there are no issues.

**Comment by damithc** [comment]
Thanks @macareonie @itstrueitstrueitsrealitsreal 

Regarding the class name, reasons why it is set as `seedu.edu.Duke` and not `Duke`:
* By the time students are supposed to merge the Gradle branch, their main class is supposed to be in a package already, and the class name should have been updated to match their Bot name too. So, setting it to `Duke` will not work for students.
* While `seedu.edu.Duke` doesn't work, it gives students a better example of how to set this property. If we set it to just `Duke`, they might think this should be set to the class name only.

**Comment by jedkohjk** [comment]
OS: Windows 11
JDK distribution: Oracle OpenJDK
Same as the above comments, `master` and `full-template` run fine, but not `add-gradle-support`, for the same reasons.

`master`:
![master](https://github.com/user-attachments/assets/ebdd26b5-8108-4cee-888e-43c9ee818691)

`full-template`:
![full-template](https://github.com/user-attachments/assets/f0ecbd68-d4d6-4be9-add0-0f3a91a1c137)

`add-gradle-support`:
![add-gradle-support](https://github.com/user-attachments/assets/9881a1f5-c1d9-4026-b4c1-4fcf002487da)

**Comment by hjungwoo01** [comment]
OS: M3 Pro (Mac Silicon)
JDK: OpenJDK version 17.0.11

I get the exact same results as @jedkohjk.

**Comment by aureliony** [comment]
OS: Windows 10
JDK: Zulu openjdk 17.0.11

Tested gradle support on add-gradle-support branch of duke-J17.

`gradlew run` throws `Error: Could not find or load main class seedu.duke.Duke.`
![idea64_nfQyd8nxoZ](https://github.com/user-attachments/assets/57f96112-3b29-4c62-9215-6444be2bb74b)

`gradlew check` and `gradlew test` ran without any errors.
`gradlew shadowJar` produces 2 different fat jar files: `duke.jar` and `duke-J17.jar`.
Running `java -jar duke.jar` produces the same error as `gradlew run`.

Running `java -jar duke-J17.jar` produces a different error:
`no main manifest attribute, in duke-J17.jar`

After moving `Duke.java` to `seedu/duke/Duke.java` and setting its package (see https://github.com/aureliony/duke-J17/commit/3a5edd7d9ee6f4936d7ef33a7191f743a57c5eb6), all gradle commands work as expected.
![idea64_AJ8Stjdkyg](https://github.com/user-attachments/assets/c5c8847a-9009-4789-80f8-1c930c6c4a0e)

However, `duke-J17.jar` still throws the error `no main manifest attribute, in duke-J17.jar`, while `duke.jar` works fine.
![explorer_A5GO894TDL](https://github.com/user-attachments/assets/79414894-fdb5-49f3-aaad-cb9e2964c99d)
![image](https://github.com/user-attachments/assets/754ee6ff-c078-415f-8439-edd2fd717b2f)

**Comment by damithc** [comment]
> However, `duke-J17.jar` still throws the error `no main manifest attribute, in duke-J17.jar`, while `duke.jar` works fine.

I think we should update the build.gradle file so that it generates only one jar file. What do you all think?

**Comment by aureliony** [comment]
> > However, `duke-J17.jar` still throws the error `no main manifest attribute, in duke-J17.jar`, while `duke.jar` works fine.
> 
> I think we should update the build.gradle file so that it generates only one jar file. What do you all think?

Yes, https://github.com/aureliony/duke-J17/commit/00dc07a91afcc2bdd3684426a363b15c52b5182b fixes the issue.

**Comment by billyhoce** [comment]
Tested on **Windows 10 Home v22H2** using **Oracle OpenJDK JDK 17.0.10**
Results are identical to that of those above

> > However, `duke-J17.jar` still throws the error `no main manifest attribute, in duke-J17.jar`, while `duke.jar` works fine.
> 
> I think we should update the build.gradle file so that it generates only one jar file. What do you all think?

Agree, and verified that @aureliony 's solution fixes the issue

**Comment by damithc** [comment]
> Yes, [aureliony@00dc07a](https://github.com/aureliony/duke-J17/commit/00dc07a91afcc2bdd3684426a363b15c52b5182b) fixes the issue.

@aureliony can send a PR for this, for both duke and AB3?

**Comment by aureliony** [comment]
Tested on AB3-J17; only 1 jar file `addressbook.jar` is generated, so no changes required for that.
![image](https://github.com/user-attachments/assets/cbe1314a-d165-43cf-95be-ba41b879a299)

**Comment by Carlintyj** [comment]
Checked out PR #2 on a MacOS (Silicon) with Azul zulu-17 version 17.0.11.

Build was successful and only produced 1 FAT JAR file.
<img width="812" alt="Screenshot 2024-07-17 at 4 09 20 AM" src="https://github.com/user-attachments/assets/8cd28656-6075-4fbe-84ee-8b2d3a459ef0">
<img width="352" alt="Screenshot 2024-07-17 at 4 09 30 AM" src="https://github.com/user-attachments/assets/7bf38465-0aef-4dd9-b2a9-31f9e7672633">

However, the same issue as stated by most above still remains.
<img width="726" alt="Screenshot 2024-07-17 at 4 11 00 AM" src="https://github.com/user-attachments/assets/d598c302-fddc-4575-af9c-bfbc28dfc2cd">

**Comment by bennyLCK** [comment]
Tested on **Windows 11 OS** using **Oracle OpenJDK Version 17.0.10 Java SDK**. 

Compiling and running the Duke.java file on **Master** branch as well as building it using Gradle after changing the target main class in build.gradle file to simply `Duke` instead of `seedu.duke.Duke` runs successfully since `Duke.java` is not in a package yet on the **Master** branch, while the alternative is to perform the changes that @aureliony proposed for `add-gradle-support` branch although there is no need since that branch would be merged into master later on.

Building the project on the **full-template** branch works perfectly fine since `Duke.java` is now in the correct directory `seedu/duke/Duke.java` and the problem about 2 JAR files being generated have been resolved by removing the `dependsOn("distZip", "distTar")` line in the `ShadowJar` segment in `build.gradle`.

Building the project in the **add-gradle-support** branch results in the same problems as already mentioned above @aureliony, regarding 2 JAR files being produced and being unable to run either of them, again resolved in @aureliony's latest PR.

In conclusion, apart from the latest PR to resolve the issue on `add-gradle-support` branch, there are likely no other problems that needs fixing and I agree with the observations above.

**Comment by teoks0199** [comment]
**OS used:** Windows 11
**JDK distribution used:** Oracle
java version "17.0.5" 2022-10-18 LTS

No issues with compiling and running ` duke.java` on **master** branch.

No issues with running JAR file on **full-template** branch.

For the **add-gradle-support** branch, same issue about 2 JAR files being generated with duke-J17.jar not being able to run. This is resolved by running `gradlew shadowJar` after performing the changes in @aureliony 's PR which will only result in 1 JAR file generated and also updating `build.gradle` to point to the correct Main class. However, running `gradlew build` still generates 2 JAR files.

**Comment by brein62** [comment]
**OS used:** M1 MacBook Air 2020 running on MacOS 13.3 Ventura
**JDK distribution:** OpenJDK Zulu 17.0.11

Results are the same as everyone else above. I agree that no changes are required for the 3 branches, except for #2.

Double-checked all branches to ensure that no traces of Java/JDK 11 were left using `grep -r 11 .`.

- The **master** branch works with IntelliJ's run interface. 
- The **full-template** branch works for all common Gradle actions, including run, shadowJar, clean, test. The shadowJar only creates 1 JAR for this branch.

For the `build.gradle` file in the **add-gradle-support** branch, seems that running `./gradlew run` immediately doesn't work, due to the `ClassNotFoundException` caused by `seedu.duke.Duke` not being found:

<img width="512" alt="image" src="https://github.com/user-attachments/assets/d25663e3-507d-428d-8ed3-93daa90300e9">

However, as mentioned above, this is intentional as students are required to organise the code into [packages](https://nus-cs2103-ay2324s1.github.io/website/admin/ip-w3.html) before and rename the class name in `build.gradle` to match their own package name. Refactoring the `Duke` class and placing it into the `seedu.duke` package works and the `./gradlew run` runs successfully. After this, `./gradlew test` and `./gradlew clean shadowJar` runs properly.

Like the rest, `./gradlew clean shadowJar` creates 2 JARs `Duke.jar` and `Duke-J17.jar`, where the first one runs properly and the second one leads to this error:

<img width="669" alt="image" src="https://github.com/user-attachments/assets/3f968c15-2281-44c2-b64d-84f9b8771e57">

Following #2 causes `./gradlew clean shadowJar` to create only `Duke.jar` which runs properly.

**Comment by baskargopinath** [comment]
M2 Pro Macbook Pro - Sonoma 14.5
IntelliJ IDEA 2022.3.2 (Community Edition

No issue with `master` and `full-template` branches

Issue with `add-gradle-support` as mentioned above due to the wrong main class:
Error can be replicated by switching to the 2nd branch and running ./gradlew run
![image](https://github.com/user-attachments/assets/4f5ce29e-27b6-47d0-917a-b3f161e19cbf)

After changing to the correct main class - `mainClass.set("Duke")` - error is resolved
![image](https://github.com/user-attachments/assets/defc262b-7afd-484f-8e9e-04327cdad70a)

Next issue is when doing `gradle build`, 2 jar files is generated 
The file i highlighted, `duke-J17.jar` doesnt work when using java -jar to launch it
![image](https://github.com/user-attachments/assets/e59a205b-a158-4a28-ac7e-a5e728a21962)


https://github.com/nus-oss/duke-J17/pull/2 will make sure 1 jar file is generated

**Comment by jyue487** [comment]
OS used: Mac Silicon
JDK distribution: Azul-Zulu 17.0.11
`master` runs perfectly fine
`full-template` runs perfectly fine
`add-gradle-support` has two issues:
    1. 2 jar files are generated when `./gradlew shadowJar` is executed. (Issue is currently handled by our people)
    2. Class Name issue. (Since prof mentioned it in [this comment](https://github.com/nus-oss/duke-J17/issues/1#issuecomment-2227375648), I created a seedu.duke package and test it, it is working perfectly fine too)

<img width="554" alt="Screenshot 2024-07-19 at 6 20 06 PM" src="https://github.com/user-attachments/assets/9b3f5109-5770-43da-b723-a356aa07dccb">

So far, no other issues are found.

**Comment by gongg21** [comment]
Tested on **M1 Pro Mac Silicon** with **Azul zulu-17 version 17.0.11**

**`master`**
- Ran Duke.java with no issues.

**`full-template`**
- Ran Duke.java with no issues.
- Executed the Gradle commands run, clean, build, test, shadowJar with no issues.
- Generated singular JAR file that runs successfully.

**`add-gradle-support`**
- Ran Duke.java with no issues.
- Executed the Gradle commands clean, build, test, shadowJar with no issues.
- Observation is similar to the rest with `gradlew run`, build fails. Changing the Main Class to 'Duke' resolves.

<img width="690" alt="Screenshot 2024-07-27 at 12 08 42 PM" src="https://github.com/user-attachments/assets/3ba8351b-abbc-49b1-a5ed-8e74ab34ae44">

- Issue with two JAR files generated from shadowJar with inability to load duke-J17.jar (same manifest error). Resolved in PR above, checked with no further issues.

<img width="202" alt="Screenshot 2024-07-27 at 12 10 01 PM" src="https://github.com/user-attachments/assets/b01f7a1c-6029-4c3c-b73a-c5c5f5580f50">
<img width="627" alt="Screenshot 2024-07-27 at 12 11 09 PM" src="https://github.com/user-attachments/assets/ff6aa22c-cb33-403e-9b1f-577274dbc244">

**Comment by drustanyjt** [comment]
OS: Ubuntu 22.04 Linux
JDK: OpenJDK 17
IDE: IntelliJ Ultimate IDEA

`master`
-  Run Duke,java works
![image](https://github.com/user-attachments/assets/a0bcf63e-2d83-40f3-9308-3a6112ffb2bf)

`full-template`
- Run Duke.java works (similar image to above)
- Gradle build, clean, test, shadowJar, run are all working.
![image](https://github.com/user-attachments/assets/2a57e68c-42a7-427c-99da-1fca1ec871d7)
- Generated jar works from cli
![image](https://github.com/user-attachments/assets/115b7e4d-25db-41e8-8c4b-2606713e9b5e)

`add-gradle-support`
- The issue mentioned above of 'gradlew run' not working still persists (as intended).
- The latest version of this branch now only has 1 jar file being generated. Which works after changing the main class to 'Duke'
![image](https://github.com/user-attachments/assets/02e93dc4-0644-4f45-9347-1ed7829441b9)

**Comment by damithc** [comment]
Thanks all for your inputs. Closing this issue now.
